Don't use context caching for gathering systemd services

terminalmage opened the pull request:
In a single Salt run, when multiple pkg states (with accompanying service states) are run, the first service state establishes ``__context__['systemd.systemd_services']``. Then, subsequent pkg states which install new packages will in effect add new systemd services, but systemd.py is oblivious to this change and further attempts to get the list of systemd services are returned from the (now outdated) cache.

This PR resolves this by not using context caching for the list of systemd services.

Fixes #40279.

terminalmage committed:
Don't use context caching for gathering systemd services

In a single Salt run, when multiple pkg states (with accompanying
service states) are run, the first service state establishes
__context__['systemd.systemd_services']. Then, subsequent pkg states
which add new packages will in effect add new systemd services, but
systemd.py is oblivious to this change and further attempts to get the
list of systemd services are returned from the (now outdated) cache.

This commit resolves this by not using context caching for the list of
systemd services.

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @thatch45, @kiorky and @techhat to be potential reviewers.

Commit:
Remove unused import for lint

terminalmage commented:
Some legit issue with unit tests, probably due to the context stuff being ripped out. Will take care of this.

terminalmage committed:
systemd.py: when getting all services, don't repeat gathering of systemd services

Now that we are not caching these in the context dunder, we should pass
them into this function so that we don't repeat gathering the systemd
services inside _get_sysv_services().

Also, this fixes an incorrect change to the mocking that was apparently
made at some point to the get_all unit test.

terminalmage commented:
OK, the bug in the unit test was, while not fatal, a sign that since we are no longer caching the systemd service list, we should be passing the systemd services into ``_get_sysv_services()`` to prevent them from being gathered a second time (so that services with both a systemd unit file and a sysv init script aren't identified as sysv). I've made that change and also fixed a (non-fatal) issue with the mocking for that test. When @cro modified systemd.py in d92e2ea1, the mocking was incorrectly updated to reflect that change.

cachedout commented:
@terminalmage Still seeing failures. Did you push up the changes you mentioned in your last comment?

terminalmage committed:
Merge branch 'issue40279' of https://github.com/terminalmage/salt into issue40279

terminalmage commented:
Yeah I remember pushing, but I didn't look at the output from the push command TBH. It looks like the push failed because you pushed a commit to my fork? I didn't even know that was possible.

terminalmage commented:
It's also a bit unsettling.

terminalmage commented:
OK, pushed and tests running.

cachedout commented:
Yeah, you can edit PRs if you have write access to the repo. It is possible for contributors to disable this ability when the PR is submitted.

terminalmage commented:
Ah, OK.

kiorky commented:
If i remember well, it will make a regression here, because we introduced caching because that was really "SLOW" without, when you have a big highstate. Can we just do better cache invalidations instead of going again on the non caching plan ?

terminalmage commented:
I think you're mistaken. The most significant performance improvement for gathering systemd services did not come from context caching. We switched from parsing the ``systemctl list-unit-files`` output and instead now use ``os.listdir()`` on the directories which contain unit files. There may have been other areas where we benefited from context caching, and it is true that eliminating it here will cause Salt to do slightly more work, but the impact should be minimal since the context dict lives for the life of a Salt run; it does not persist in the minion's memory.

If you have a reliable solution for getting unrelated (i.e. non-``service``) states to invalidate systemd's context cache, I'm all ears, but since we can't control or anticipate all the ways in which other states or remote execution function calls will affect the available systemd services, I don't see a bulletproof solution. We could make all ``pkg`` virtual functions which use ``__context__`` clear the systemd services cache when there are changes to the installed packages, but there are other ways of adding services, such as using a ``file.managed`` state to add a unit file, or even generating it using ``cmd.run``. However, playing whack-a-mole with these edge cases and corner cases seems much more of a hassle than simply running a few extra calls to ``os.listdir()``.

damon-atkins commented:
Every thought of calling dbus and systemd from python?
https://github.com/systemd/python-systemd
https://dbus.freedesktop.org/doc/dbus-python/doc/tutorial.html
Might be able to even have dbus watches like file watches
